### DISCUSSION TITLE
Ability to edit chat history for context optimization

### DISCUSSION THREAD

**Discussion by versality** [olimorris/codecompanion.nvim#398] [Ideas]
# Chat History Management

First of all, excellent work on the plugin! After testing various Neovim options, this is arguably the best solution currently available.

## Feature Request: Chat History Editing
I'd like to suggest a feature that I personally find missing - the ability to edit chat history. Often during longer conversations with the LLM, I want to:
- Remove irrelevant parts of previous exchanges
- Keep the meaningful context
- Continue the conversation from arbitrary point

This is similar to the "branching" feature available in both OpenAI and Anthropic UIs, where users can edit previous messages and create new conversation branches.

## Current Behavior
At first, since intuitively it felt possible, I have modified existing chat history, but after checking debug mode, it's clear that modifying submitted messages doesn't update the internal memory. From reviewing the code, it seems message modification only occurs when removing the system prompt by adjusting `messages[1]`.

## Implementation Challenges
I understand this feature would require significant code changes:
- Rebuilding the entire messages dataset from chat history
- Implementing clear identifiers/separators for each chat message in the Neovim buffer
- Handling potential edge cases

@olimorris, have you considered implementing this feature? Would love to hear your thoughts.

**Comment by olimorris** [comment]
Thanks for such a thoughtful and well written enhancement.

This is something that has been on my mind for sometime too. I'd like to at least make `gd` editable. I also have plans to improve `gr` which regenerates the last response from the LLM.

Previously, you could alter historic messages and Tree-sitter would re-parse the whole buffer but when it came to adding slash commands and variables it would have been cumbersome trying to workout where they should fit into the message stack. Given my assumed lower priority of such a feature I went ahead with just parsing the last message and no one complained.

I've got a pretty full roadmap for the next few weeks but will give this some thought as to how we could add this in.

**Comment by olimorris** [comment]
PR #753 is ready for testing and the docs are fully updated. You can edit message history now via the debug window (`gd`)
